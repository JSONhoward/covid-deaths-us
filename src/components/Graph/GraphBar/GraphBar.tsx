import React, { CSSProperties } from 'react'
import { Transition } from 'react-transition-group';
import anime from 'animejs'
import Tooltip from '@material-ui/core/Tooltip'

import { GraphBarStyled, Bar, Label, Deaths, fadeDuration } from './GraphBar.styled';

type Props = {
    height: number,
    visible: boolean,
    option: string,
    data: InitialResponseType,
}

const tooltipStyles = {
    container: { display: 'flex', flexDirection: 'column', justifyContent: 'center', height: '125px', width: '250px' },
    rows: { display: 'flex', justifyContent: 'space-between', fontWeight: 'bold', fontSize: '.75rem', width: '100%' }
}

const GraphBar: React.FC<Props> = ({ height, visible, option, data }) => {
    const [animDeaths, setAnimDeaths] = React.useState(0)

    React.useEffect(() => {
        let valObj = {
            value: 0
        }

        anime({
            targets: valObj,
            value: data.covid_19_deaths,
            round: 1,
            easing: 'easeInOutQuad',
            duration: 2000,
            autoplay: false,
            loop: false,
            update: () => {
                setAnimDeaths(valObj.value)
            }
        }).play()
    }, [data.covid_19_deaths])

    return (
        <Transition in={visible} timeout={fadeDuration}>
            {
                state => (
                    <GraphBarStyled state={state}>
                        <Label>
                            {option === 'state' ? data.state : (data.age_group_new === '85' ? '85+' : data.age_group_new === '1' ? 'under 1' : data.age_group_new)}
                        </Label>
                        <Tooltip title={
                            <React.Fragment>
                                <div style={tooltipStyles.container as CSSProperties}>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Total:</p><p>{data.total_deaths}</p>
                                    </div>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Pneumonia, influenza, or Covid-19:</p><p>{data.pneumonia_influenza_or_covid}</p>
                                    </div>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Covid-19:</p><p>{data.covid_19_deaths}</p>
                                    </div>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Pneumonia:</p><p>{data.pneumonia_deaths}</p>
                                    </div>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Pneumonia and Covid-19:</p><p>{data.pneumonia_and_covid_19_deaths}</p>
                                    </div>
                                    <div style={tooltipStyles.rows as CSSProperties}>
                                        <p>Influenza:</p><p>{data.influenza_deaths}</p>
                                    </div>
                                </div>
                            </React.Fragment>
                        }
                            placement='right-start'>
                            <Bar barHeight={height} state={state} />
                        </Tooltip>
                        <Deaths>
                            {animDeaths}
                        </Deaths>
                    </GraphBarStyled>
                )
            }
        </Transition>
    )
}

export default GraphBar
