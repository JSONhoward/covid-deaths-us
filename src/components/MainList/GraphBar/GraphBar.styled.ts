import styled, {keyframes} from 'styled-components'

export const fadeDuration = 1000

const redBarAnimation = keyframes`
0%{background-position:92% 0%}
50%{background-position:9% 100%}
100%{background-position:92% 0%}
`

export const GraphBarStyled = styled('div')<{state: string}>`
display: flex;
flex-direction: column-reverse;
align-items: flex-end;
height: 100%;
width: calc(90vw / 5);
min-width: 75px;
padding: 5px;
opacity: ${props => {
    switch(props.state) {
        case 'entering':
            return 1
        case 'entered':
            return 1
        case 'exiting':
            return 0
        case 'exited':
            return 0
        default:
            return 1
    }
}};
transition: opacity ${fadeDuration}ms ease-in-out;
`

export const Bar = styled('div')<{barHeight: number, state: string}>`
height: ${props => {
    switch(props.state) {
        case 'entering':
            return props.barHeight + 'px'
        case 'entered':
            return props.barHeight + 'px'
        case 'exiting':
            return 0
        case 'exited':
            return 0
        default:
            return props.barHeight + 'px'
    }
}};
width: 100%;
background: linear-gradient(220deg, #ff0000, #b91616);
background-size: 600% 600%;

-webkit-animation: ${redBarAnimation} 7s ease infinite;
-moz-animation: ${redBarAnimation} 7s ease infinite;
animation: ${redBarAnimation} 7s ease infinite;
transition: height 2s ease-in-out;
`

export const Label = styled('div')`
display: flex;
height: 2rem;
width: 100%;
justify-content: center;
overflow: hidden;
`

export const Deaths = styled('p')`
position: relative;
top: -1rem;
height: 1rem;
width: 100%;
text-align: center;
font-weight: bold;
transform: rotate(-45deg);
`