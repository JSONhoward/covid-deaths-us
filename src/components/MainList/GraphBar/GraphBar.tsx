import React from 'react'
import { Transition } from 'react-transition-group';
import anime from 'animejs'

import { GraphBarStyled, Bar, Label, Deaths, fadeDuration } from './GraphBar.styled';

type Props = {
    height: number,
    stateName: string,
    deaths: number,
    visible: boolean,
    age: string,
    options: string
}

const GraphBar: React.FC<Props> = ({ height, stateName, deaths, visible, age, options }) => {
    const [animDeaths, setAnimDeaths] = React.useState(0)

    React.useEffect(() => {
        let valObj = {
            value: 0
        }

        anime({
            targets: valObj,
            value: deaths,
            round: 1,
            easing: 'easeInOutQuad',
            duration: 2000,
            autoplay: false,
            loop: false,
            update: () => {
                setAnimDeaths(valObj.value)
            }
        }).play()
    }, [deaths])

    return (
        <Transition in={visible} timeout={fadeDuration}>
            {
                state => (
                    <GraphBarStyled state={state}>
                        <Label>
                            {options === 'state' ? stateName : age}
                        </Label>
                        <Bar barHeight={height} state={state} />
                        <Deaths>
                            {animDeaths}
                        </Deaths>
                    </GraphBarStyled>
                )
            }
        </Transition>
    )
}

export default GraphBar
