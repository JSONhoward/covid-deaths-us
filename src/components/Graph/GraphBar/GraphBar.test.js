import React from 'react'
import {create} from 'react-test-renderer'
import {screen, render} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import GraphBar from './'

describe('GraphBar Tests', () => {
    test('Snapshot Test', () => {
        const height = 1000,
        visible = true,
        option = 'test option',
        data = {
            "data_as_of": "2020-09-16T00:00:00.000",
            "start_week": "2020-02-01T00:00:00.000",
            "end_week": "2020-09-12T00:00:00.000",
            "state": "New York City",
            "sex": "All Sexes",
            "age_group_new": "All Ages",
            "covid_19_deaths": "20686",
            "total_deaths": "59803",
            "pneumonia_deaths": "11055",
            "pneumonia_and_covid_19_deaths": "7947",
            "influenza_deaths": "967",
            "pneumonia_influenza_or_covid": "23928"
          }

        const tree = create(<GraphBar height={height} visible={visible} option={option} data={data} />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})