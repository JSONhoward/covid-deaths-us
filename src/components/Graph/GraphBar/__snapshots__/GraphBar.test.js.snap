// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphBar Tests Snapshot Test 1`] = `
<div
  className="sc-bdnylx ktDbDz"
>
  <div
    className="sc-dlnjPT jCUDoH"
  >
    All Ages
  </div>
  <div
    aria-describedby={null}
    className="sc-gtssRu hNqDcK"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title={null}
  />
  <p
    className="sc-hKFyIo jTqqGX"
  >
    0
  </p>
</div>
`;
